File: Estruturais/Adapter/Composite/Composite01.js
class Componente{
    constructor(nome){
        this.nome = nome;
    }

    adicionar(){}

    remover(){}


    obterNome(){}

}

class Folha extends Componente{
    constructor(nome, preco){
        super(nome);
        this.preco = preco;
    }

    obterNome(){
        return this.nome;
    }

    obterPreco(){
        return this.preco;
    }
}

class Conteiner extends Componente{
    constructor(nome){
        super(nome);
        this.componentes = [];
    }
    adicionar(componentes){
        this.componentes.push(componentes);
    }
    remover(componentes){
        const index = this.componentes.indexOf(componentes);
        this.componentes.splice(index, 1)
    }
    obterNome(){
        return this.nome;
    }

    obterPreco(){
        let preco = 0;
        this.componentes.forEach(componentes => {
            preco += componentes.obterPreco();
        });

        return preco;
    }
}

// usando  o recurso como fruta e preco ////
const maca = new Folha("Maça", 2.50);
const laranja = new Folha("Lanja", 3.00);
const uva = new Folha("Uva", 5.00)

const frutas = new Conteiner("Frutas");


frutas.adicionar(maca);
frutas.adicionar(laranja);
frutas.adicionar(uva);


const caixa = new Conteiner("Caixa de frutas");
caixa.adicionar(frutas);
caixa.adicionar(new Folha("Embalagem", 1.00));
caixa.adicionar(new Folha("Emabalagem", 1.00));
caixa.adicionar(new Folha("Embalagem",1.00));


console.log (`Preço Total: R$${caixa.obterPreco()}`);

File: Estruturais/Adapter/Composite/Composite01.py
class Componente:
    def __init__(self,nome):
        self.nome = nome
     
    def  adicionar(self):
        pass
    
    def  remover(self):
        pass
    
    def  obterNome(self):
        pass
    
    def  obterPreco(self):
        pass
    
class Folha(Componente):
    def __init__(self, nome, preco):
        super().__init__(nome)
        self.preco = preco
        
    def obterNome(self):
        return self.nome
    
    def obterPreco(self):
        return self.preco
    
    
class Conteiner(Componente):
    def __init__(self, nome):
        super().__init__(nome)
        self.componentes = []

    def adicionar(self,componente):
        self.componentes.append(componente)
        
    def remover(self,componente):   
        index = self.componentes.ondex(componente)
        self.componentes.pop(index)
        
    def obterNome(self):
        return self.nome 
    
    def obterPreco(self):
        preco = 0
        for componente in self.componentes:
            preco += componente.obterPreco()
        return preco
            
            
            
# Exemplo de uso 

maca = Folha('Maça', 2.50)
laranja = Folha('Laranja', 3.0)
uva = Folha('Uva' , 5.0)

frutas = Conteiner("Frutas")
frutas.adicionar(maca)
frutas.adicionar(laranja)
frutas.adicionar(uva)


caixa = Conteiner("Caixa de Frutas")
caixa.adicionar(frutas)
caixa.adicionar(Folha('Embalagem', 1.0))


print(f"Preço total: R${caixa.obterPreco()}")
        